modules/mssql/variables.tf

```diff
@@ -237,11 +237,8 @@ variable "user_password" {
 
 variable "additional_users" {
   description = "A list of users to be created in your cluster"
-  type = list(object({
-    name     = string
-    password = string
-  }))
-  default = []
+  type        = list(map(any))
+  default     = []
```

Comment:
We had received some feedback that not having vars typed hurt UX as users were unable to determine what values to pass in. WDYT about switching to use [optionals](https://developer.hashicorp.com/terraform/language/expressions/type-constraints#optional-object-type-attributes) instead? This will be breaking change requiring 1.3.0 but we expect most modules to switch to use optionals next year.

Reply:
I don't think we should force all our customers to upgrade terraform to 1.3.0+, for a small fearure fix. I think we should wait and only cause this breaking change when it is absolutely necessary, wdyt?

As an alternate solution we can have one more resource called "additional_users_with_random_password", where users can set null or "" to the password field to generate random password (replicating the behavior of postgres module) in mssql module. In this way we are not breaking anything for the exisiting mssql customers, what do you think? Should I go forward with this?

Reply:
We will likely switch to 1.3 in the next breaking release as we had multiple breaking major releases ([12.0](https://github.com/terraform-google-modules/terraform-google-sql-db/pull/329),[11.0](https://github.com/terraform-google-modules/terraform-google-sql-db/pull/288),[10.0](https://github.com/terraform-google-modules/terraform-google-sql-db/pull/277),[8.0](https://github.com/terraform-google-modules/terraform-google-sql-db/pull/251)) due to changes in object types which 1.3 will avoid. Agreed we don't need to force the switch for this feature, I was thinking of bundling this with next major if we decided to use optionals.

With the alternate approach using `additional_users_with_random_password`, would we need a new input to target this? Also re reading the description, I suspect most users may not want to use a literal `""` since no password is supported by `null`. It might just be cleaner to not support literal `""`.

Reply:
Setting password as "", creates an user with an empty password, then that user can login without providing password.

1) Should I use Optionals and send the PR, where you can merge this PR once we decide to create a breaking release. or
2)  With additional_users_with_random_password, yes, it would be a new input field for the customer. Now mssql customers could either use additonal_users or additonal_users_with_random_password input fields for creating users. The mssql customers can use the new input if they require their users to be set with an random password (and they still would have the option to change the random password if needed. This new field would replicate the behaviour of additional_users field in postgres module).

Should I go ahead with (1) or (2) ?

Reply:
> If password is set as null, then no password is created for the user, and if the password is set as "", then the password is also literally "" (empty password).
> >Setting password as "", creates an user with an empty password, then that user can login without providing password. 

In that case, what is the difference between empty string `""` as password vs no password at all via `null`? 

I am wondering even with optionals we will need to provide a default value signifying when to generate the random password - would we use empty string or null?

Reply:
In postgres module, random password is generated during empty string and null values for password, but for MySQL, random password is generated only when password field is not set in the additional_users resource.
```
additional_users([{name: "test-user"}])
```
I think we can also use optionals in mssql to make the password variable optional, so that the mssql users would also have an option for not setting the password, so that a random password is only generated when the password field is not set. If we generate random password for a null or an empty string password for a user, then it would be a breaking change, since some mssql customers would be setting the password as null or "" to create  a login user without pwd.